File: RetrofitReservation/app/src/main/java/ma/ensa/retrofitreservation/ListClientActivity.java
package ma.ensa.retrofitreservation;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import ma.ensa.retrofitreservation.adapters.ClientAdapter;
import ma.ensa.retrofitreservation.models.Client;
import ma.ensa.retrofitreservation.network.RetrofitInstance;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ListClientActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ClientAdapter adapter;
    private List<Client> clients = new ArrayList<>();
    private Button buttonAddClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list_client);

        recyclerView = findViewById(R.id.list_clients);
        buttonAddClient = findViewById(R.id.button_add_client);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        // Initialize adapter with empty data
        adapter = new ClientAdapter(clients, this);
        recyclerView.setAdapter(adapter);

        // Fetch all clients and measure latency
        getAllClients();

        // Set up swipe-to-delete functionality
        setUpSwipeToDelete();

        buttonAddClient.setOnClickListener(v -> {
            Intent intent = new Intent(ListClientActivity.this, AddClientActivity.class);
            startActivity(intent);
        });
    }

    private void getAllClients() {
        long startTime = System.currentTimeMillis();  // Capture start time

        Call<List<Client>> call = RetrofitInstance.getApi().getAllClients();
        call.enqueue(new Callback<List<Client>>() {
            @Override
            public void onResponse(Call<List<Client>> call, Response<List<Client>> response) {
                long endTime = System.currentTimeMillis();  // Capture end time
                long latency = endTime - startTime;  // Calculate latency
                Log.d("NetworkLatency", "GET Request Latency: " + latency + " ms");

                if (response.isSuccessful()) {
                    List<Client> fetchedClients = response.body();
                    if (fetchedClients != null && !fetchedClients.isEmpty()) {
                        clients.clear();
                        clients.addAll(fetchedClients);
                        adapter.notifyDataSetChanged();
                    } else {
                        Toast.makeText(ListClientActivity.this, "Aucun client trouvé.", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(ListClientActivity.this, "Erreur lors de la récupération des clients.", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<Client>> call, Throwable t) {
                Toast.makeText(ListClientActivity.this, "Erreur : " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void deleteClient(Long clientId, int position) {
        long startTime = System.currentTimeMillis();  // Capture start time

        Call<Void> call = RetrofitInstance.getApi().deleteClient(clientId);
        call.enqueue(new Callback<Void>() {
            @Override
            public void onResponse(Call<Void> call, Response<Void> response) {
                long endTime = System.currentTimeMillis();  // Capture end time
                long latency = endTime - startTime;  // Calculate latency
                Log.d("NetworkLatency", "DELETE Request Latency: " + latency + " ms");

                if (response.isSuccessful()) {
                    clients.remove(position);
                    adapter.notifyItemRemoved(position);
                    Toast.makeText(ListClientActivity.this, "Client supprimé", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(ListClientActivity.this, "Erreur lors de la suppression", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Void> call, Throwable t) {
                Toast.makeText(ListClientActivity.this, "Erreur : " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void setUpSwipeToDelete() {
        ItemTouchHelper.SimpleCallback simpleCallback = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                int position = viewHolder.getAdapterPosition();
                Client client = clients.get(position);

                new AlertDialog.Builder(ListClientActivity.this)
                        .setMessage("Voulez-vous vraiment supprimer ce client ?")
                        .setCancelable(false)
                        .setPositiveButton("Oui", (dialog, which) -> {
                            if (client.getId() != null) {
                                deleteClient(client.getId(), position);
                            }
                        })
                        .setNegativeButton("Non", (dialog, which) -> adapter.notifyItemChanged(position))
                        .create()
                        .show();
            }
        };

        new ItemTouchHelper(simpleCallback).attachToRecyclerView(recyclerView);
    }
}vv

File: ReservationRetrofit_KOTLIN/app/src/main/java/ma/ensaj/reservationretrofit_kotlin/ListClientActivity.kt
package ma.ensaj.reservationretrofit_kotlin

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import ma.ensaj.reservationretrofit_kotlin.adapters.ClientAdapter
import ma.ensaj.reservationretrofit_kotlin.models.Client
import ma.ensaj.reservationretrofit_kotlin.network.RetrofitInstance
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okio.Buffer
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ListClientActivity : AppCompatActivity() {

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: ClientAdapter
    private val clients: MutableList<Client> = mutableListOf()
    private lateinit var buttonAddClient: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list_client)

        recyclerView = findViewById(R.id.list_clients)
        buttonAddClient = findViewById(R.id.button_add_client)

        recyclerView.layoutManager = LinearLayoutManager(this)

        // Initialize adapter with empty data
        adapter = ClientAdapter(clients)
        recyclerView.adapter = adapter

        // Measure latency for fetching clients
        measureLatencyForFetchingClients()

        // Set up swipe-to-delete functionality
        setUpSwipeToDelete()

        buttonAddClient.setOnClickListener {
            val intent = Intent(this@ListClientActivity, AddClientActivity::class.java)
            startActivity(intent)
        }
    }

    private fun measureLatencyForFetchingClients() {
        val startTime = System.currentTimeMillis()

        // Perform network call to fetch all clients
        val call = RetrofitInstance.getApi().getAllClients()
        call.enqueue(object : Callback<List<Client>> {
            override fun onResponse(call: Call<List<Client>>, response: Response<List<Client>>) {
                val latency = System.currentTimeMillis() - startTime

                // Calculate the data size in KB
                val dataSizeInKB = calculateResponseSizeInKB(response)
                Log.d("ListClientActivity", "Latency for GET (fetch clients): $latency ms")
                Log.d("ListClientActivity", "Data size for GET (fetch clients): $dataSizeInKB KB")

                if (response.isSuccessful) {
                    val fetchedClients = response.body()
                    Log.d("ListClientActivity", "Fetched clients: $fetchedClients")
                    if (!fetchedClients.isNullOrEmpty()) {
                        clients.clear()
                        clients.addAll(fetchedClients)
                        adapter.notifyDataSetChanged() // Notify adapter to refresh the data
                    } else {
                        Toast.makeText(this@ListClientActivity, "Aucun client trouvé.", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    Toast.makeText(this@ListClientActivity, "Erreur lors de la récupération des clients.", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<List<Client>>, t: Throwable) {
                Toast.makeText(this@ListClientActivity, "Erreur : ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun measureLatencyForDelete(clientId: Long, position: Int) {
        val startTime = System.currentTimeMillis()

        // Proceed with deleting the client
        deleteClient(clientId, position, startTime)
    }

    private fun deleteClient(clientId: Long, position: Int, startTime: Long) {
        // In many cases, DELETE requests don't include a body, just the ID in the URL
        val call = RetrofitInstance.getApi().deleteClient(clientId)

        // Log the data size for the DELETE request (optional, if sending data)
        val deleteDataSizeInKB = calculateRequestSizeInKB(clientId)
        Log.d("ListClientActivity", "Data size for DELETE (delete client): $deleteDataSizeInKB KB")

        call.enqueue(object : Callback<Void> {
            override fun onResponse(call: Call<Void>, response: Response<Void>) {
                val latency = System.currentTimeMillis() - startTime
                Log.d("ListClientActivity", "Latency for DELETE (delete client): $latency ms")

                if (response.isSuccessful) {
                    clients.removeAt(position)
                    adapter.notifyItemRemoved(position)
                    Toast.makeText(this@ListClientActivity, "Client supprimé", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this@ListClientActivity, "Erreur lors de la suppression", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Void>, t: Throwable) {
                Toast.makeText(this@ListClientActivity, "Erreur : ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun calculateResponseSizeInKB(response: Response<List<Client>>): Double {
        // Convert the response body to bytes and then to kilobytes
        val responseBody = response.body()?.let {
            val jsonString = RetrofitInstance.getGson().toJson(it)
            val byteArray = jsonString.toByteArray(Charsets.UTF_8)
            byteArray.size.toDouble()
        } ?: 0.0

        return responseBody // Convert bytes to kilobytes
    }

    private fun calculateRequestSizeInKB(clientId: Long): Double {
        // Prepare a JSON-like string for the DELETE request (just the ID)
        val jsonString = "{\"id\": $clientId}"
        val byteArray = jsonString.toByteArray(Charsets.UTF_8)

        return byteArray.size.toDouble()
    }

    private fun setUpSwipeToDelete() {
        val simpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                val client = clients[position]

                android.app.AlertDialog.Builder(this@ListClientActivity)
                    .setMessage("Voulez-vous vraiment supprimer ce client ?")
                    .setCancelable(false)
                    .setPositiveButton("Oui") { _, _ ->
                        client.id?.let { measureLatencyForDelete(it, position) }
                    }
                    .setNegativeButton("Non") { _, _ ->
                        adapter.notifyItemChanged(position) // Restore the item
                    }
                    .create()
                    .show()
            }
        }

        ItemTouchHelper(simpleCallback).attachToRecyclerView(recyclerView)
    }
}
